src/controller/python/chip/native/PyChipError.h

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPError.h>
+
+// Note: We can use a constructor here to avoid "ToPyChipError", but it will result in compilers warning for incompatible with C
+// linkage.
+struct PyChipError
+{
+    uint32_t mCode;
+    uint32_t mLine;
+    const char * mFile;
+};
+
+// To avoid the error message "user-defined type 'PyChipError' which is incompatible with C"
+
+inline bool operator!(const PyChipError & aError)
+{
+    return aError.mCode != 0;
+}
+
+inline bool operator==(const PyChipError & aError, const CHIP_ERROR & aOther)
+{
+    return aError.mCode == aOther.AsInteger();
+}
+
+inline bool operator==(const CHIP_ERROR & aOther, const PyChipError & aError)
+{
+    return aError.mCode == aOther.AsInteger();
+}
+
+inline bool operator!=(const PyChipError & aError, const CHIP_ERROR & aOther)
+{
+    return !(aError.mCode == aOther.AsInteger());
+}
+
+inline bool operator!=(const CHIP_ERROR & aOther, const PyChipError & aError)
+{
+    return !(aError.mCode == aOther.AsInteger());
+}
+
+inline PyChipError ToPyChipError(const CHIP_ERROR & aError)
+{
+    return PyChipError
+    {
+        .mCode = aError.AsInteger(),
```

Comment:
Could we store along with it the SDK formatted `AsString()` equivalent? That would leverage all the formatters installed for each of the different SDK parts, and make it easier to debug.

That could then be displayed in Python.

Reply:
I would suggest not to store such string as I think the error type should be relative small, and it should not contain malloc-ed pointers (so the Python side can safely ignore the errors when they want).

The current PyChipError is trivially copiable (at least logically).

I added another format function, I also think we should update the ChipError class to support formatting error into a user provided buffer.

Reply:
`pychip_FormatError` achieves what I want of leveraging the SDK's formatter, while avoiding the malloc/free issues. Thank you!